### ISSUE TITLE
Warn (or tidy) poorly formed boolean attributes

### ISSUE THREAD

**Issue by brennanyoung** [htacg/tidy-html5#796]
Given that [the only valid value for a boolean attribute](https://www.w3.org/TR/html5/infrastructure.html#sec-boolean-attributes) is the name of the attribute or empty string, it would be useful if tidy would warn about cases such as:

`hidden="false"`

At first glance this looks meaningful, but `hidden="false"` means the same as 
`hidden="hidden"`, `hidden=""` and `hidden="true"` which is not at all intuitive.

It should be better known that the value of a boolean attribute is always treated as true if present, regardless of the attribute value. If our markup-tidying tools spotted those kinds of mistakes, it would be useful.

Here's a (possibly complete?) list of boolean attributes in html5:
```
allowFullscreen
async
autofocus
autoplay
checked
compact
controls
declare
default
defaultChecked
defaultMuted
defaultSelected
defer
disabled
draggable
enabled
formNoValidate
hidden
indeterminate
inert
isMap
itemScope
loop
multiple
muted
noHref
noResize
noShade
noValidate
noWrap
open
pauseOnExit
readOnly
required
reversed
scoped
seamless
selected
sortable
spellcheck
translate
trueSpeed
typeMustMatch
visible
```

**Comment by geoffmcl** [comment]
@brennanyoung this seems a laudable aim - in brief - `to warn/fix on boolean attributes`... thanks for the issue... and W3C ref. link...

While you have given a snippet example, `hidden="false"`, thanks, can you please put it in context...

That is, supply one, or **more**, minimum html input examples, tidy version under test, config used, current output, and **expected** output...

I quickly wrote one, `input5\in_796.html` -

```html
<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Is. #796</title>
</head>
<body>
<div hidden="false">
<p>Bad boolean attribute</p>
</div>
</body>
</html>
```

Tidy `next` **5.7.20** outputs - `No warnings or errors were found.`... Hmmmm... **not** good...

The Nu Html Checker flags it -

```
Error: Bad value false for attribute hidden on element div.
From line 8, column 1; to line 8, column 20
d>↩<body>↩<div hidden="false">↩<p>Ba
```

But need **more** test inputs, minimum, examples... thanks...

Concerning your list of `boolean attributes in html5`, that would need to be compared to the internal `libTidy` [attr.c](https://github.com/htacg/tidy-html5/blob/next/src/attrs.c#L99) ... look for `CH_BOOL`, in the static table... saw some matches...

Explored with `--strict-tags-attributes yes` and `--literal-attributes yes`, but no effect... maybe there are other options to try... need testing...

Searched issues for `boolean attribute`, and found #58, #69, #142, #395, #436, ... particularly, open #439, #654, ... others?... which need to be explored, reread, for any overlap/pending...

Otherwise, looks a valid tidy `enhancement`, but do not have such a label, so using FR...

Feedback, comments, patches, and eventually agreed PR, very welcome... thanks...

**Comment by brennanyoung** [comment]
Good. Here are another couple of example inputs.

### Incorrect boolean attribute value literally "false":
```
<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Is. #796</title>
</head>
<body>
<video controls="false" width="250" src="/media/examples/flower.mp4"></video>
<p>Boolean attribute test</p>
</body>
</html>
```

### Incorrect boolean attribute value implicitly"false":
```
<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Is. #796</title>
</head>
<body>
<p visible="hidden">Boolean attribute test</p>
</body>
</html>
```

### Rationale for these two examples:
It would be difficult to deduce whether a poorly formed attribute value is intended to be 'false'. Example: visible="hidden". I suppose the correct behaviour would be to *remove* the attribute entirely in the cases where it has the literal value "false" (I think that would be a 'fair' intervention), but otherwise to warn if it is anything other than empty string or a case insensitive match for the attribute name. 

The two examples provided here could be used to distinguish the two possible outputs.

**Comment by brennanyoung** [comment]
Yes, I should have been clearer about that. <p visible="hidden"> is indeed invalid. I include this example because it 'looks' right (the intended value 'false' is strongly implied, even though the markup is poorly formed). Tidy should (and does) catch this.

**Comment by geoffmcl** [comment]
@brennanyoung thanks for the clarification... but I think we should, initially at least, stick with samples that demonstrate just **this** issue... 

Continuing the code research...

I added a sample with `<input ... Checked="True">`, since `checked` is already marked as `CH_BOOL`, thus will go through `CheckBool`... to see what happens...

As suspected, although, while in `CheckLowerCaseAttrValue`, the `T` is **silently** changed to `t`, ie lowercase - why? - it makes no check that the `value` **matches** the attribute `name`... 

Note, in `CheckBool`, `Checked` had already been converted to `checked`? Need to check, understand, where that happens... why... etc...

Then the code to **match** would be simple to **add**...

Like the `tidy` equivalent of `if (strcmp(attval->attribute,attval->value)) { printf("TODO: Msg, we have a BA prob.\n"); }`... need to add the correct report/message/enum... 

And then to decide what, if anything, to do about it... ie any fix? and adjusting the chosen warning `message` to this...

Your initial comment showed some 44 boolean attributes... is that **all**? ... need to check, verify W3C references... but that aside...

However `attrs.c` has only some 21 marked as `CH_BOOL`... 

So even if the above pseudo code addition were implemented, only some 50% of the attributes, the `BA`, would pass through it... this needs to be corrected, changed, modified....

If, like `hidden` and `control`, changing them from a `NULL` to `CH_BOOL` seems an easy ok, but each of the other changes needs to be checked... verified... understood... one by one...

So seem to have identified some **step** to achieving this goal... which I will work towards...

Any comments, feedback, comparing lists/tables, code, patches, etc, etc... very welcome... thanks...

**Comment by linusjf** [comment]
Any updates on this? As far as I can see, tidy does not have any options to abbreviate Boolean html attributes to their most concise form, just the attribute name. Is there going to be an option for this, any time soon? Are there any opinionated html formatters that provide this feature? At the moment, if tidy encounters an abbreviated Boolean attribute, it fills it , rather than leaving it be. Why should tidy do that?
